core/src/main/java/org/infinispan/reactive/publisher/impl/PartitionAwareClusterPublisherManager.java

```diff
@@ -128,24 +129,38 @@ public void start() {
    }
 
    private <R> SegmentCompletionPublisher<R> registerPublisher(SegmentCompletionPublisher<R> original) {
-      return (subscriber, segmentsComplete) -> {
-         // Processor has to be serialized due to possibly invoking onError from a different thread
-         FlowableProcessor<R> earlyTerminatingProcessor = AsyncProcessor.<R>create().toSerialized();
-         pendingProcessors.add(earlyTerminatingProcessor);
-
-         // Have to check after registering in case if we got a partition between when the publisher was created and
-         // subscribed to
-         if (isPartitionDegraded()) {
-            pendingProcessors.remove(earlyTerminatingProcessor);
-            earlyTerminatingProcessor.onError(CLUSTER.partitionDegraded());
-            original.subscribe(earlyTerminatingProcessor);
-         } else {
-            Flowable<R> actualPublisher = Flowable.<R>fromPublisher(s -> original.subscribe(s, segmentsComplete))
-                  .doFinally(earlyTerminatingProcessor::onComplete);
-            Publisher<R> flowableAsPublisher = earlyTerminatingProcessor
-                  .doFinally(() -> pendingProcessors.remove(earlyTerminatingProcessor));
-            Flowable.merge(flowableAsPublisher, actualPublisher)
-                  .subscribe(subscriber);
+      return new SegmentCompletionPublisher<R>() {
+         @Override
+         public void subscribeWithSegments(Subscriber<? super Notification<R>> subscriber) {
+            handleEarlyTermination(subscriber, scp -> Flowable.<Notification<R>>fromPublisher(scp::subscribeWithSegments));
+         }
+
+         @Override
+         public void subscribe(Subscriber<? super R> subscriber) {
+            handleEarlyTermination(subscriber, Flowable::fromPublisher);
+         }
+
+         private <S> void handleEarlyTermination(Subscriber<? super S> subscriber, Function<SegmentCompletionPublisher<R>, Flowable<S>> function) {
```

Comment:
This is another where the listener is in case if something happens in between values. We don't want to allow the status to change to degraded then back again and not cause an issue. There is no telling what values we may have missed at that point. Since the original subscriber may be filtering or collapsing whole values.

Reply:
Why would the status going to degraded and back be a problem? 
If all partitions got degraded and the segment ownership did not change, then the values didn't change. 
If another partition did stay available and allowed modifications, then this node became a non-owner, at least for a while, and the `SegmentAwarePublisherImpl` check would catch the change.

Reply:
But it would catch the change by saying the segment is suspect, not that it was suspect because of degradation. I have not thought through the various scenarios of partition degradation as you, so I would recommend if you think this change is okay to put it in a separate PR as I am not comfortable changing this behavior myself.

Reply:
If going degraded and back is indeed a problem, we can also use the simpler CompletableFuture-based notification

```java
            CompletableFuture<Void> cf = new CompletableFuture<>();
            pendingCompletableFutures.add(cf);

            Flowable.fromPublisher(function.apply(original))
                           .doOnNext(s -> throwExceptionIfNeeded(cf))
                           .doFinally(() -> pendingProcessors.remove(cf))

...
   private void throwExceptionIfNeeded(CompletableFuture<Void> cf) {
      if (cf.isDone())
         cf.join();
   }

```

Reply:
> If going degraded and back is indeed a problem, we can also use the simpler CompletableFuture-based notification
> 
I don't know if it is a problem, that is why I am hesitant changing this logic in a PR that isn't related to partition handling. My changes here are merely the API changes.

Reply:
But I trust you if you say it should be okay :)

Reply:
I'm pushing on all of these because I also have a WIP that makes cache sets use publishers for everything internally, removing the need for each interceptor to implement a `CacheSet` wrapper. So I'm interested in making the publisher-based stuff just as fast as the current stream-based stuff :)